packages/near-api-js/src/providers/wallet-rpc-provider.ts

```diff
@@ -0,0 +1,473 @@
+            changes_type: 'account_changes',
+            account_ids: accountIdArray,
+            block_id: blockId,
+            finality
+        });
+    }
+
+    /**
+     * Gets contract state changes for a given array of accountIds
+     * pass block_id OR finality as blockQuery, not both
+     * Note: If you pass a keyPrefix it must be base64 encoded
+     * See [docs for more info](https://docs.near.org/docs/develop/front-end/rpc#view-contract-state-changes)
+     * @returns {Promise<ChangeResult>}
+     */
+    async contractStateChanges(accountIdArray: string[], blockQuery: BlockReference, keyPrefix = ''): Promise<ChangeResult> {
+        const { finality } = blockQuery as any;
+        const { blockId } = blockQuery as any;
+        return this.sendJsonRpc('EXPERIMENTAL_changes', {
+            changes_type: 'data_changes',
+            account_ids: accountIdArray,
+            key_prefix_base64: keyPrefix,
+            block_id: blockId,
+            finality
+        });
+    }
+
+    /**
+     * Gets contract code changes for a given array of accountIds
+     * pass block_id OR finality as blockQuery, not both
+     * Note: Change is returned in a base64 encoded WASM file
+     * See [docs for more info](https://docs.near.org/docs/develop/front-end/rpc#view-contract-code-changes)
+     * @returns {Promise<ChangeResult>}
+     */
+    async contractCodeChanges(accountIdArray: string[], blockQuery: BlockReference): Promise<ChangeResult> {
+        const { finality } = blockQuery as any;
+        const { blockId } = blockQuery as any;
+        return this.sendJsonRpc('EXPERIMENTAL_changes', {
+            changes_type: 'contract_code_changes',
+            account_ids: accountIdArray,
+            block_id: blockId,
+            finality
+        });
+    }
+
+    /**
+     * Returns gas price for a specific block_height or block_hash.
+     * @see [https://docs.near.org/api/rpc/gas](https://docs.near.org/api/rpc/gas)
+     *
+     * @param blockId Block hash or height, or null for latest.
+     */
+    async gasPrice(blockId: BlockId | null): Promise<GasPrice> {
+        return await this.sendJsonRpc('gas_price', [blockId]);
+    }
+
+    /**
+     * Directly call the RPC specifying the method and params
+     *
+     * @param method RPC method
+     * @param params Parameters to the method
+     */
+    async sendJsonRpc<T>(method: string, params: object): Promise<T> {
+        const response = await exponentialBackoff(REQUEST_RETRY_WAIT, REQUEST_RETRY_NUMBER, REQUEST_RETRY_WAIT_BACKOFF, async () => {
+            try {
+                if (this._provider) {
+                    const request: RequestParams = {
+                        method,
+                        params,
+                        id: (_nextId++),
+                        jsonrpc: '2.0'
+                    };
+                    const response = await this._provider.request(request);
+                    if (Array.isArray(response)) {
```

Comment:
Under what conditions is `response` an array? Is this behavior that `this._provider.request` is required to implement?

Reply:
The transaction hash value is returned as an array of strings when a user sends a transaction. Because for wallets that support multi-transaction, the hash of the transaction is returned as an array. But `near-api-js` support only sending one transaction, so we use `response[0]`.